R/api.R

```diff
@@ -19,9 +19,51 @@ build_api <- function(runner, path, backup_period = NULL, rate_limit = 2 * 60) {
   backup <- orderly_backup(runner$config, backup_period)
   api$registerHook("preroute", backup$check_backup)
   api$registerHook("preroute", check_timeout(runner, rate_limit))
+  api$registerHook("preroute", api_preroute)
+  api$registerHook("postserialize", api_postserialize)
   api
 }
 
+api_preroute <- function(data, req, res, value) {
+  api_log_start(data, req, res)
+}
+
+api_postserialize <- function(data, req, res, value) {
+  api_log_end(data, req, res, value)
+}
+
+api_log_start <- function(data, req, res) {
+  api_log(sprintf("%s %s", req$REQUEST_METHOD, req$PATH_INFO))
+}
+
+api_log_end <- function(data, req, res, value) {
+  if (is.raw(res$body)) {
+    size <- length(res$body)
+  } else {
+    size <- nchar(res$body)
+  }
+  if (res$status >= 400 &&
+      identical(res$headers[["Content-Type"]], "application/json")) {
+    dat <- jsonlite::parse_json(res$body)
+    for (e in dat$errors) {
+      if (!is.null(e$key)) {
+        api_log(sprintf("error-key: %s", e$key))
+        api_log(sprintf("error-detail: %s", e$detail))
+        if (!is.null(e$trace)) {
+          trace <- sub("\n", " ", vcapply(e$trace, identity))
+          api_log(sprintf("error-trace: %s", trace))
```

Comment:
codecov is flagging this, and I think it's worth triggering this to be sure we're getting useful debug out

Reply:
Thanks - that picked up that we don't have a `key` here. That is a thing hintr is adding.